src/mcmc/gibbs.jl

```diff
@@ -620,10 +709,14 @@ end
 """
 Run a Gibbs step for the first varname/sampler/state tuple, and recursively call the same
 function on the tail, until there are no more samplers left.
+
+The `step_function` argument should always be either AbstractMCMC.step or
+AbstractMCMC.step_warmup.
 """
 function gibbs_step_recursive(
     rng::Random.AbstractRNG,
     model::DynamicPPL.Model,
+    step_function::Function,
```

Comment:
would this create some kind of type instability? my first order thought is that `Function` is abstract type

Reply:
I think it should be fine, since each function has its own type. E.g.
```julia
julia> typeof(identity)
typeof(identity) (singleton type of function identity, subtype of Function)

julia> typeof(sin)
typeof(sin) (singleton type of function sin, subtype of Function)
```
The `::Function` bit just enforces that you can't pass as the `step_function` argument anything that isn't of type `Function`, but the compiler will still see the concrete type of the argument. I also checked that making this change didn't have a substantial impact on runtime.